<a href="https://colab.research.google.com/github/biswajit017/Python_Pandas/blob/main/Day2_Pandas_Series_part_2_Supplementary.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Some Important Series Methods

In [ ]:
# astype
# between
# clip
# drop_duplicates
# isnull
# dropna
# fillna
# isin
# apply
# copy

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
subs = pd.read_csv('/content/subs.csv').squeeze()
subs

,Subscribers gained
0,48
1,57
2,40
3,43
4,44
...,...
360,231
361,226
362,155
363,144


In [ ]:
vk = pd.read_csv('/content/kohli_ipl.csv',index_col='match_no').squeeze()
vk

,runs
match_no,
1,1
2,23
3,13
4,12
5,1
...,...
211,0
212,20
213,73


In [ ]:
movies = pd.read_csv('/content/bollywood.csv',index_col='movie').squeeze()
movies

,lead
movie,
Uri: The Surgical Strike,Vicky Kaushal
Battalion 609,Vicky Ahuja
The Accidental Prime Minister (film),Anupam Kher
Why Cheat India,Emraan Hashmi
Evening Shadows,Mona Ambegaonkar
...,...
Hum Tumhare Hain Sanam,Shah Rukh Khan
Aankhen (2002 film),Amitabh Bachchan
Saathiya (film),Vivek Oberoi


In [ ]:
# astype

import sys
sys.getsizeof(vk)

3472

astype() is used to convert datatype, often to reduce memory usage and improve performance.


In [ ]:
sys.getsizeof(vk.astype('int16'))

2182

In [ ]:
# between

vk[vk.between(51,99)].size

43

between() in pandas is used to check whether values lie within a given range.

It returns True for values inside the range and False otherwise.

Here in the above example 51 and 99 both are included

In [ ]:
# clip
subs


,Subscribers gained
0,48
1,57
2,40
3,43
4,44
...,...
360,231
361,226
362,155
363,144


In [ ]:
subs.clip(100,200)

,Subscribers gained
0,100
1,100
2,100
3,100
4,100
...,...
360,200
361,200
362,155
363,144


clip() in pandas is used to limit values within a specified range.

Values smaller than the minimum are replaced by the minimum, and values larger than the maximum are replaced by the maximum.

On the above example 144 is come under the 100 and 200 it will remain same.

In [ ]:
# drop_duplicates

temp = pd.Series([1,1,2,2,3,3,4,4])
temp

,0
0,1
1,1
2,2
3,2
4,3
5,3
6,4
7,4


drop_duplicates() in pandas is used to remove duplicate rows from a DataFrame or Series.

By default, it keeps the first occurrence and removes repeated values.

In [ ]:
temp.drop_duplicates()

,0
0,1
2,2
4,3
6,4


It also gives additional flexibiity that is we can keep second repeated value and delete the first repeated value but bydefault it is first

In [ ]:
temp.drop_duplicates(keep='last')

,0
1,1
3,2
5,3
7,4


Removes duplicate values/rows from temp and keeps the last occurrence of each duplicate.

In [ ]:
movies.drop_duplicates()

,lead
movie,
Uri: The Surgical Strike,Vicky Kaushal
Battalion 609,Vicky Ahuja
The Accidental Prime Minister (film),Anupam Kher
Why Cheat India,Emraan Hashmi
Evening Shadows,Mona Ambegaonkar
...,...
Sssshhh...,Tanishaa Mukerji
Rules: Pyaar Ka Superhit Formula,Tanuja
Right Here Right Now (film),Ankit


In [25]:
temp.duplicated()

,0
0,False
1,True
2,False
3,True
4,False
5,True
6,False
7,True


duplicated() in pandas is used to identify duplicate rows or values present in a series or not.

It returns True for duplicate entries and False for unique entries.

In [33]:
print(temp.duplicated().sum())

4


In [35]:
print(vk.duplicated().sum())

137


In [36]:
temp = pd.Series([1,2,3,np.nan,5,6,np.nan,8,np.nan,10])
temp

,0
0,1.0
1,2.0
2,3.0
3,NaN
4,5.0
5,6.0
6,NaN
7,8.0
8,NaN
9,10.0


In [37]:
temp.size

10

In [39]:
print(temp.count())

7


isnull

In [43]:
print(temp.isnull().sum())

3


# dropna()

dropna() does not change the original data by default.

It creates and returns a new Series/DataFrame with null values removed, and stores the result there if assigned.

the datatype often becomes float.
Reason

Missing values (NaN) in pandas are stored as floating-point values.
So integer columns containing NaN are automatically converted to float.

In [44]:
temp.dropna()

,0
0,1.0
1,2.0
2,3.0
4,5.0
5,6.0
7,8.0
9,10.0


#fillna()
Suppose i dont want to drop those missing values then i can use fillna() function and replace those none value with some numbers or zero.

the datatype often becomes float.
Reason

Missing values (NaN) in pandas are stored as floating-point values.
So integer columns containing NaN are automatically converted to float.

In [45]:
temp.fillna(0)

,0
0,1.0
1,2.0
2,3.0
3,0.0
4,5.0
5,6.0
6,0.0
7,8.0
8,0.0
9,10.0


In [47]:
temp.fillna(33)

,0
0,1.0
1,2.0
2,3.0
3,33.0
4,5.0
5,6.0
6,33.0
7,8.0
8,33.0
9,10.0


If i want to replace those none value with mean of the dataset then i will write

In [48]:
temp.fillna(temp.mean())

,0
0,1.0
1,2.0
2,3.0
3,5.0
4,5.0
5,6.0
6,5.0
7,8.0
8,5.0
9,10.0


# isin()

isin() in pandas is used to check whether values are present in a given list, set, or collection.

It returns True for matching values and False otherwise.

In [51]:
vk[(vk == 49) | (vk == 99)]  # Without using the function

,runs
match_no,
82,99
86,49


In [52]:
vk[vk.isin([49,99])]

,runs
match_no,
82,99
86,49


# apply()

apply() is a Pandas function used to apply a custom function to each element of a Series or to rows/columns of a DataFrame for performing operations and transformations.

Criteria  :
Now down below i want the lead actor first name with uppercase  


In [54]:
movies

,lead
movie,
Uri: The Surgical Strike,Vicky Kaushal
Battalion 609,Vicky Ahuja
The Accidental Prime Minister (film),Anupam Kher
Why Cheat India,Emraan Hashmi
Evening Shadows,Mona Ambegaonkar
...,...
Hum Tumhare Hain Sanam,Shah Rukh Khan
Aankhen (2002 film),Amitabh Bachchan
Saathiya (film),Vivek Oberoi


In [56]:
movies.apply(lambda x:x.split())

,lead
movie,
Uri: The Surgical Strike,"[Vicky, Kaushal]"
Battalion 609,"[Vicky, Ahuja]"
The Accidental Prime Minister (film),"[Anupam, Kher]"
Why Cheat India,"[Emraan, Hashmi]"
Evening Shadows,"[Mona, Ambegaonkar]"
...,...
Hum Tumhare Hain Sanam,"[Shah, Rukh, Khan]"
Aankhen (2002 film),"[Amitabh, Bachchan]"
Saathiya (film),"[Vivek, Oberoi]"


In [58]:
movies.apply(lambda x:x.split()[0].upper())

,lead
movie,
Uri: The Surgical Strike,VICKY
Battalion 609,VICKY
The Accidental Prime Minister (film),ANUPAM
Why Cheat India,EMRAAN
Evening Shadows,MONA
...,...
Hum Tumhare Hain Sanam,SHAH
Aankhen (2002 film),AMITABH
Saathiya (film),VIVEK


In [59]:
subs

,Subscribers gained
0,48
1,57
2,40
3,43
4,44
...,...
360,231
361,226
362,155
363,144


Here from subs dataset i want mean of the dataset then want to calculate that if it is greater than mean then good day otherwise bad day

In [60]:
subs.apply(lambda x:'good day' if x > subs.mean() else 'bad day')

,Subscribers gained
0,bad day
1,bad day
2,bad day
3,bad day
4,bad day
...,...
360,good day
361,good day
362,good day
363,good day


# copy()

copy() is used in Pandas to create a separate independent copy of a DataFrame or Series, so changes made to the copied object do not affect the original data.

In [70]:
vk

,runs
match_no,
1,100
2,23
3,13
4,12
5,1
...,...
211,0
212,20
213,73


In [63]:
new = vk.head()

# wehen we are using head or tail() they will give us view of the data not copy or replica so whatever we will update it will directly update in the original dataset

In [69]:
new

,runs
match_no,
1,100
2,23
3,13
4,12
5,1


In [68]:
new[1] = 100

head() may return a view of the original data. Therefore, modifying the returned object can sometimes update the original dataset.
on the above situation it is updated because prevously in vk, vk[1] was 1 after updated to 100 new one also changed as well as vk also changed
  so

To avoid this, use .copy().


In [72]:
new = vk.head().copy()


In [73]:
new[1] = 200

In [74]:
new

,runs
match_no,
1,200
2,23
3,13
4,12
5,1


In [75]:
vk

,runs
match_no,
1,100
2,23
3,13
4,12
5,1
...,...
211,0
212,20
213,73
